### PR TITLE
Remove type for size arg on Streamer constructor

### LINKED ISSUE #12422: Can't download large folders (~>1GB) - Error "Argument 2 passed to OC\Streamer::__construct() must be of the type integer, float given..."
### Steps to reproduce
1. Try to download a large folder (>1GB).
2. Error is thrown while trying to generate a zip file.

### Expected behaviour
Zip file should be generated correctly and downloaded correctly.

### Actual behaviour
An error is thrown to the user with a reqId to request support.

## Server configuration detail

**Operating system:** Linux 4.9.0-8-686-pae #1 SMP Debian 4.9.130-2 (2018-10-27) i686

**Webserver:** nginx/1.10.3 (fpm-fcgi)

**Database:** mysql 10.1.26

**PHP version:** 

7.0.30-0+deb9u1
Modules loaded: Core, date, libxml, openssl, pcre, zlib, filter, hash, Reflection, SPL, session, standard, cgi-fcgi, mysqlnd, PDO, xml, apcu, apc, calendar, ctype, curl, dom, mbstring, fileinfo, ftp, gd, gettext, iconv, igbinary, imagick, imap, intl, json, exif, mcrypt, mysqli, pdo_mysql, pdo_sqlite, Phar, posix, readline, redis, shmop, SimpleXML, sockets, sqlite3, sysvmsg, sysvsem, sysvshm, tokenizer, wddx, xmlreader, xmlwriter, xsl, zip, Zend OPcache

**Nextcloud version:** 14.0.3 - 14.0.3.0

**Updated from an older Nextcloud/ownCloud or fresh install:** 

**Where did you install Nextcloud from:** Nextcloud official web

<details><summary>Signing status</summary>

</details>

<details><summary>List of activated apps</summary>

```
Enabled:
 - accessibility: 1.0.1
 - activity: 2.7.0
 - calendar: 1.6.3
 - cloud_federation_api: 0.0.1
 - comments: 1.4.0
 - contacts: 2.1.6
 - dav: 1.6.0
 - federatedfilesharing: 1.4.0
 - federation: 1.4.0
 - files: 1.9.0
 - files_external: 1.5.0
 - files_pdfviewer: 1.3.2
 - files_sharing: 1.6.2
 - files_texteditor: 2.6.0
 - files_trashbin: 1.4.1
 - files_versions: 1.7.1
 - files_videoplayer: 1.3.0
 - firstrunwizard: 2.3.0
 - gallery: 18.1.0
 - issuetemplate: 0.4.0
 - logreader: 2.0.0
 - lookup_server_connector: 1.2.0
 - nextcloud_announcements: 1.3.0
 - notes: 2.4.2
 - notifications: 2.2.1
 - oauth2: 1.2.1
 - password_policy: 1.4.0
 - provisioning_api: 1.4.0
 - serverinfo: 1.4.0
 - sharebymail: 1.4.0
 - support: 1.0.0
 - survey_client: 1.2.0
 - systemtags: 1.4.0
 - tasks: 0.9.7
 - theming: 1.5.0
 - twofactor_backupcodes: 1.3.1
 - updatenotification: 1.4.1
 - workflowengine: 1.4.0
Disabled:
 - admin_audit
 - encryption
 - user_external
 - user_ldap

```
</details>

<details><summary>Configuration (config/config.php)</summary>

```
{
    "instanceid": "***REMOVED SENSITIVE VALUE***",
    "passwordsalt": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trusted_domains": [
        "***REMOVED SENSITIVE VALUE***"
    ],
    "datadirectory": "***REMOVED SENSITIVE VALUE***",
    "overwrite.cli.url": "***REMOVED SENSITIVE VALUE***",
    "dbtype": "mysql",
    "version": "14.0.3.0",
    "dbname": "***REMOVED SENSITIVE VALUE***",
    "dbhost": "***REMOVED SENSITIVE VALUE***",
    "dbtableprefix": "oc_",
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "logtimezone": "UTC",
    "installed": true,
    "loglevel": 0,
    "memcache.local": "\\OC\\Memcache\\APCu",
    "maintenance": false,
    "theme": "",
    "mail_smtpmode": "smtp",
    "mail_smtpauthtype": "PLAIN",
    "mail_smtpsecure": "tls",
    "mail_from_address": "***REMOVED SENSITIVE VALUE***",
    "mail_domain": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpauth": 1,
    "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpport": "587",
    "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
    "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
    "updater.secret": "***REMOVED SENSITIVE VALUE***"
}
```
</details>

**Are you using external storage, if yes which one:** No

**Are you using encryption:** No 

**Are you using an external user-backend, if yes which one:** No

## Client configuration

**Browser:** Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36

**Operating system:** Mac OS X Mojave 10.14.1

## Logs

<details><summary>Browser log</summary>

```

```
</details>

<details><summary>Nextcloud log</summary>

```
TypeError: Argument 2 passed to OC\Streamer::__construct() must be of the type integer, float given, called in /var/www/html/owncloud/lib/private/legacy/files.php on line 166
/var/www/html/owncloud/lib/private/legacy/files.php - line 166:

OC\Streamer->__construct(OC\AppFramework\Http\Request {}, 12288299774, 893)

/var/www/html/owncloud/apps/files/ajax/download.php - line 64:

OC_Files::get("__SENSITIVE_INFO_HIDDEN__", "__SENSITIVE_INFO_HIDDEN", { head: false})

/var/www/html/owncloud/lib/private/Route/Route.php - line 155:

require_once("/var/www/ht ... p")


OC\Route\Route->OC\Route\{closure}("*** sensiti ... *")

/var/www/html/owncloud/lib/private/Route/Router.php - line 297:

call_user_func(Closure {}, { _route: "f ... "})

/var/www/html/owncloud/lib/base.php - line 987:

OC\Route\Router->match("/apps/files/ajax/download.php")

/var/www/html/owncloud/index.php - line 42:

OC::handleRequest()
```
</details>

<details><summary>Browser log</summary>

</details>

I have tried with several folders, but I always get the same problem. Notice that is a 32-bit server with latest Debian version.

### LINKED ISSUE #15117: TypeError: Argument 2 passed to OC\Streamer::__construct() must be of the type integer, float given, called in /config/www/nextcloud/lib/private/legacy/files.php on line 166
### Steps to reproduce
1. Try to download 2 files (file 1 is 2,8GB, file 2 just a few KB)  at the same time via share link or logged in as admin.
2. Throws error right away.
3. Does not throw error when 1 file is downloaded at a time. System does not zip the file.
4. Problem does not occur when downloading a smaller (1GB) file together with another file, system creates a ZIP and files are downloaded.

I am wondering if it is possible to turn off the ZIP function when multiple files are downloaded, and whether that would resolve the issue.

### Expected behaviour
Files should both be downloaded

### Actual behaviour
Error is thrown. 

### Server configuration
Alpine 3.9 in Docker

**Web server:**
NGINX

**Database:**
MariaDB

**PHP version:**
7.2.13

**Nextcloud version:** (see Nextcloud admin page)
15.0.6

**Updated from an older Nextcloud/ownCloud or fresh install:**
Fresh install

**Where did you install Nextcloud from:**
https://hub.docker.com/r/linuxserver/nextcloud/

### PR DESCRIPTION
**Pull request by artonge**
Resolves #12422 and #15117

On 32 bit systems integers can not be bigger than 2 147 483 647. This is problematic as it throws an error when downloading a folder bigger than 2Go because php will cast the size to float.

### CODE REVIEW

**Comment by scriptator** [comment]
Can somebody review and merge this? For me it seems fine and fixes the problem which I experienced

**Review by nickvergessen — APPROVED** [review]
Urgh, while I can see this solving the problem you have, I guess the issue will pop in quite a lot other locations too. I just think that at some point we just can not manage to stay compatible with 16-bit anymore.

**Comment by kesselb** [comment]
> the issue will pop in quite a lot other locations too.

Yes. Another report for a similar problem is: https://github.com/nextcloud/server/issues/13160. Unfortunately https://wiki.php.net/rfc/bigint is still work in progress.

A) Add a library like https://github.com/brick/math and use BigInteger for size attributes.
B) Don't type hint `$size`

I don't like A nor B :see_no_evil: But there are many raspi instances out there.

**Comment by nickvergessen** [comment]
Well, you forgot:
C) Don't support mutli GB files on those rapsi instances :see_no_evil:

**Comment by scriptator** [comment]
> Urgh, while I can see this solving the problem you have, I guess the issue will pop in quite a lot other locations too.

Totally agree! It would be certainly better to fix the problem at the root

> I just think that at some point we just can not manage to stay compatible with 16-bit anymore.

What does this have to do with 16 bit? IMO raspi is running 32 bit which also makes sense when you consider the 2gb limit

**Comment by kesselb** [comment]
> Totally agree! It would be certainly better to fix the problem at the root

Where is the root? PHP not having BigInt? Raspian not using 64 bit? We don't have much options. Change the type of $size for every occurence to this BigInteger library will take months. Not adding type hints for $size is the only way I can think of.

> What does this have to do with 16 bit? IMO raspi is running 32 bit which also makes sense when you consider the 2gb limit

Typo ...

**Review by kesselb — APPROVED** [review]

**Comment by scriptator** [comment]
There is of course also another option: officially drop support for 32 bit systems. As I just found out there are 64 bit OS available for the raspberry and even the Raspberrypi foundation is now working on a 64 bit raspbian kernel. If it is clearly stated everywhere that 32 bit does not work properly and raspbian users should install a 64 bit distro I think it should also be okay

**Comment by artonge** [comment]
I think dropping support for 32-bit systems might be a bit precipitated. 64-bit CPU on a Raspberrypi only came with the version 4, so there will still be a lot of 32-bit rpi for a some years. And the rpi is not the only 32-bit single board computer that is used as a small home server.

So dropping 32-bit support would mean dropping support for a lot of self-hosters.

**Comment by kesselb** [comment]
cc @rullzer :ping_pong:

**Comment by nachoparker** [comment]
Raspbian images are still 32 bits userland, even those with a 64 bits kernel.

**Comment by kesselb** [comment]
Sorry! I forgot the milestone :see_no_evil:

File: lib/private/Streamer.php [hunk 1]
```diff
@@ -43,7 +43,7 @@ class Streamer {
 	 * @param int $numberOfFiles The number of files (and directories) that will
 	 *        be included in the streamed file
 	 */
-	public function __construct(IRequest $request, int $size, int $numberOfFiles){
+	public function __construct(IRequest $request, $size, int $numberOfFiles){
```
Comment (kesselb) [inline lib/private/Streamer.php]: We should probably add a short note to the docblock why typehinting `$size` is bad.

**Comment by go2sh** [comment]
I dont think you can change the type to float at all. I had a similar discussions  with Morris and the problem is, that you cannot represent every integer with a float exactly especially for big numbers. You might get a value that is some where around the actual size creating a completely new bunch of problems. Out of bound reads, corrupted files, changing files sizes, plus the behavior might be total random as it depends on the actual file size. Because of that, a size in nextcloud should never be float with the consequences for 32 bit instances.

**Comment by kesselb** [comment]
> size creating a completely new bunch of problems. Out of bound reads, corrupted files, changing files sizes, plus the behavior might be total random as it depends on the actual file size. 

Probably. But this pr is not about making $size a float. It's about removing the type hint.

**Comment by go2sh** [comment]
Hitting the type hint is just a symptom. The file size is never stored or transferred as float, so somewhere up the call chain, a conversion to float happens, which shouldn't happen. I can imagine only two sources, either the db or a string conversion. Finding the conversion and throwing an exception is the correct fix.

**Comment by kesselb** [comment]
PHP will cast the value from filesystem or database to float. Your "correct fix" will break it even more for 32bit users. 

```
php -r "var_dump(PHP_INT_MAX);"
int(9223372036854775807)

php -r "var_dump(9223372036854775807);"                               
int(9223372036854775807)

php -r "var_dump(9223372036854775808);"                               
float(9.2233720368548E+18)
```

**Comment by go2sh** [comment]
I traced down the problem: The `Streamer` class is called from the `ZipResponse` class. You can add an abitrary amount of resources to the response. The size of each response will be summed up. Then php kicks in: [Integer Overflow](https://www.php.net/manual/de/language.types.integer.php#language.types.integer.overflow) When a php integer overflows, it gets convered to float. :man_facepalming: So you will lose precision no matter what.

**Review by go2sh — CHANGES_REQUESTED** [review]
As per comment.

**Comment by rullzer** [comment]
Yep, we need to think of a better way to tackle this.
As removing it will also blow up in several cases as the resulting zip file will be invalid.

**Comment by artonge** [comment]
Hi,

Maybe the real solution would be to display an error message to the user explaining why the file or zipped folder can't be downloaded.

Could an exception be thrown here, and caught in an error handler to display the correct error message ?

**Comment by skjnldsv** [comment]
Ping! :)

**Comment by skjnldsv** [comment]
> Yep, we need to think of a better way to tackle this.
> As removing it will also blow up in several cases as the resulting zip file will be invalid.

What shall we do then?

**Comment by Wolfgang1966** [comment]
How about using tar archives für everything above 4GB? It is already there depending on the client os and it will at least allow to download big archives at all. Ok, Windows does not support tar archives directly, but there are enough tools out there which do.  And it is better than being not able to download the file at all.

Tried it with a 9GB archive, was immediately asked to search for a application in the MS app store which offers several of them.

Consider this a workaround, but at least it is (in my opinion) a fair improvement compared to the current situation.

**Comment by Wolfgang1966** [comment]
> How about using tar archives für everything above 4GB? It is already there depending on the client os and it will at least allow to download big archives at all. Ok, Windows does not support tar archives directly, but there are enough tools out there which do. And it is better than being not able to download the file at all.
> 
> Tried it with a 9GB archive, was immediately asked to search for a application in the MS app store which offers several of them.
> 
> Consider this a workaround, but at least it is (in my opinion) a fair improvement compared to the current situation.

My version of the constructor in Streamer.php now is:

```
                if ($size > 0 && $size < 4 * 1000 * 1000 * 1000 && $numberOfFiles < 65536) {
                        $this->streamerInstance = new ZipStreamer(['zip64' => false]);
                } else {
                        $this->streamerInstance = new TarStreamer();
                }
```

**Comment by go2sh** [comment]
@Wolfgang1966 The problem here is, that the size as float doesn't represent the actual size of the files, as floating point numbers (32bit) are only an aproximation of integers (32bit).

**Comment by Wolfgang1966** [comment]
> @Wolfgang1966 The problem here is, that the size as float doesn't represent the actual size of the files, as floating point numbers (32bit) are only an aproximation of integers (32bit).

This is true, but in that case not relevant. As you can see, the parameter is only used to distuinguish between the usage of 32bit ZIP or anything else. 32bit zip can be used until 4GiB, which is roughly 4.29 GB. As we compare the value to 4.000.000.000 4GB, there is more than enough safety margin for any kind of missing accurancy that might be introduced by using float here.

**Comment by kesselb** [comment]
`public function __construct(IRequest $request, int $size, int $numberOfFiles) {`

This code fails if you are on a 32bit operating system and size is bigger than 2147483648 because a bigger value is casted to float and then does not match the type hint anymore and :boom: 

Whatever you are talking about is a different story and unrelated to this pull request.

**Comment by MichaIng** [comment]
Btw, Raspberry Pi 2 v1.2 is already ARMv8 capable: https://en.wikipedia.org/wiki/Raspberry_Pi#Specifications
So relevant 32-bit only SBCs are the RPi Zero models only. But the Raspberry Pi OS 64-bit version is still in beta, so dropping 32-bit support would still affect most RPi users and of course it would also affect a lot of other SBCs, though with a significantly lower user number.

Not representative, but the top 5 SBCs of DietPi users participating our survey program:
RPi 4 Model B (aarch64) | 2588 | 18%
-- | -- | --
RPi 4 Model B (armv7l) | 1856 | 13%
RPi 3 Model B (armv7l) | 1325 | 9%
RPi Zero W (armv6l) | 1042 | 7%
RPi 3 Model B+ (armv7l) | 890 | 6%

`armv7l` means 32-bit OS version, which is still the majority due to limitations of the 64-bit RPi userland and GPU accelerated software builds.

**Comment by MichaIng** [comment]
Linking the very same issue present with the trash bin: #13160
Possible solution is the same.

**Comment by szaimen** [comment]
Closing due to https://github.com/nextcloud/documentation/pull/9071

**Pull request closed without merging**